posthog/cloud_utils.py

```diff
@@ -38,3 +44,45 @@ def is_cloud():
 def TEST_clear_cloud_cache(value: Optional[bool] = None):
     global is_cloud_cached
     is_cloud_cached = value
+
+
+def get_cached_instance_license() -> Optional["License"]:
+    """Returns the first valid license and caches the value for the lifetime of the instance, as it is not expected to change.
+    If there is no valid license, it returns None.
+    """
+    global instance_license_cached
+    global is_instance_licensed_cached
+
+    try:
+        from ee.models.license import License
+
+        if isinstance(instance_license_cached, License):
```

Comment:
This will fail before `instance_license_cached` is set. We need to check `'instance_license_cached' in globals()`. That way we can also avoid having `is_instance_licensed_cached`, since whether `instance_license_cached` is defined is enough of an indication (no matter if `instance_license_cached` is an `Instance` or `None`)

Reply:
I thought about combining the two globals but it's possible that that instance _isn't_ licensed (eg self-hosted) and in that case the cached license will be None, but we still don't want to check it every time to see if there is a license. So we need to know if it's licensed, and if it is, what the license is. 

> This will fail before instance_license_cached is set. We need to check `'instance_license_cached' in globals()`

it's defined at the top of the file, won't that be evaluated and set in globals before the function ever runs?

Reply:
Ah, I was definitely wrong on `This will fail before `instance_license_cached` is set.`, I missed that it's defined at the top of the file.

> in that case the cached license will be None, but we still don't want to check it every time to see if there is a license

This still holds, that's what I meant – if we _don't_ set `instance_license_cached ` at the top of the file, we won't need `is_instance_licensed_cached ` because `'instance_license_cached' in globals()` will suffice for checking whether the license or its absence has been cached.

Reply:
We could do it that way but I feel like it's a bit of an anti-pattern, I feel like globals should be defined at the top of the file / outside functions. If you feel strongly I can change the implementation but if you don't feel strongly I think I'd just keep it the way I have it.